packages/core/src/routes/session.ts

```diff
@@ -264,13 +264,27 @@ export default function sessionRoutes<T extends AnonymousRouter>(router: T, prov
 
   router.post('/session/consent', async (ctx, next) => {
     const interaction = await provider.interactionDetails(ctx.req, ctx.res);
-    const { session, grantId, params, prompt } = interaction;
+    const {
+      session,
+      grantId,
+      params: { client_id },
+      prompt,
+    } = interaction;
     assertThat(session, 'session.not_found');
 
+    console.warn('interaction', interaction);
+
     const { accountId } = session;
     const grant =
       conditional(grantId && (await provider.Grant.find(grantId))) ??
-      new provider.Grant({ accountId, clientId: String(params.client_id) });
+      new provider.Grant({ accountId, clientId: String(client_id) });
+
+    const { applicationId } = await findUserById(accountId);
+
+    if (!applicationId) {
+      // Save application id which the user first consented
+      await updateUserById(accountId, { applicationId: String(client_id) });
+    }
```

Comment:
this will be lagging under large concurrency. will be better to do this without blocking the request. please add an issue for this and put it under eng quality phase 3

Reply:
i can do it another way: add an extra middleware

```ts
oidc.use(async (ctx, next) {
    // we can reuse `ctx.oidc.entities.Account` object when user is requesting `/consent`
    await next();
});
```

we will discuss later in the next discussion.